### PR TITLE
Added calendar for executions

### PR DESCRIPTION
**Pull request by Cheshiriks**
Added calendar for executions

Снимок4432

### CODE REVIEW

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/externals/calendar/ReactCalendarBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import org.w3c.dom.events.Event
+import react.ChildrenBuilder
+import react.react
+import kotlin.js.Date
+
+/**
+ * @param onChange
+ * @param handler
+ */
+fun ChildrenBuilder.calendar(
```
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/externals/calendar/ReactCalendarBuilder.kt:14]: We should have English calendar

[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/externals/calendar/ReactCalendarBuilder.kt:14]: Added :)

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt [hunk 2]
```diff
@@ -98,30 +103,46 @@ class ExecutionController(private val executionService: ExecutionService,
...
-                    it.toDto()
-                }
-                    .reversed()
+        .zipWhen { organization ->
+            projectService.findWithPermissionByNameAndOrganization(authentication, projectName, organization.name, Permission.READ)
+        }
+        .map { (organization, _) ->
+            filters?.let {
+                val startTime = createLocalDateTime(filters.startTime)
+                val endTime = filters.endTime?.let { createLocalDateTime(it) } ?: getEndLocalDateTime(filters.startTime)
```
Comment (petertrr) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt:130]: I'd suggest to put this logic into `getEndLocalDateTime` completely

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/filters/ExecutionFilters.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Aff filters in one property
+ * @property startTime to filter by [startTime] execution formatting: <year, month, day>
+ * @property endTime to use filter by [endTime] execution formatting: <year, month, day>
+ */
+@Serializable
+data class ExecutionFilters(
+
+    val startTime: Triple<Int, Int, Int>,
```
Comment (petertrr) [inline save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/filters/ExecutionFilters.kt:13]: Please create a data class instead of `Triple`. Tuples with unnamed fields work well only when they represent local variables; with `Triple` as a class property it will become harder to keep track if all fields are unpacked correctly among the code and also harder to extend these fields, should there be a neccessity.

Also, there is `kotlinx.datetime.LocalDateTime` that can be used in both JS and JVM. Why have you decided not to use it?

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/HistoryView.kt [hunk 4]
```diff
@@ -32,10 +34,15 @@ import react.dom.html.ReactHTML.div
 import react.dom.html.ReactHTML.td
 import react.table.columns
 
+import kotlin.js.Date
```
Comment (petertrr) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/HistoryView.kt:37]: We can use `kotlinx.datetime.LocalDateTime` to have a more universal solution among platforms

File: save-frontend/build.gradle.kts [hunk 5]
```diff
@@ -49,6 +49,9 @@ kotlin {
                     }
                 }
             }
+            commonWebpackConfig {
+                cssSupport.enabled = true
```
Comment (petertrr) [inline save-frontend/build.gradle.kts:53]: Why is this required?

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/LocalDateTimeUtils.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * @param data: year, month, day
+ * @return [data] in LocalDateTime format
+ */
+fun createLocalDateTime(data: Triple<Int, Int, Int>): LocalDateTime {
+    val year = data.first
+    val month = data.second + 1
+    val day = data.third
+    return LocalDateTime.of(year, month, day, 0, 0, 0)
```
Comment (petertrr) [inline save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/LocalDateTimeUtils.kt:17]: You can use `LocalDate` instead of `LocalDateTime` if you care only about date

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/HistoryView.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/HistoryView.kt:37]: Yes, I agree. But React elements require kotlin.js.Date

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/HistoryView.kt [hunk 7]
```diff
@@ -85,19 +92,35 @@ external interface HistoryViewState : State {
...
+    /**
+     * All filters in one value [filters]
+     */
+    var filters: ExecutionFilters
+}
+
+/**
+ * [Props] of a data table with filters for execution
+ */
+external interface FiltersProps<D : Any> : TableProps<D> {
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/HistoryView.kt:105]: If it expects `ExecutionFilters`, maybe `D` should be Execution\ExecutionDto?

File: save-frontend/build.gradle.kts (hunk 5 above)
Comment (Cheshiriks) [reply save-frontend/build.gradle.kts:53]: This is necessary for external CSS style to work. 
It is called with _kotlinext.js.require("example.css")_

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/HistoryView.kt (hunk 7 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/HistoryView.kt]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt]: Removed this logic

File: save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/filters/ExecutionFilters.kt (hunk 3 above)
Comment (Cheshiriks) [reply save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/filters/ExecutionFilters.kt]: Fixed

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/LocalDateTimeUtils.kt (hunk 6 above)
Comment (Cheshiriks) [reply save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/LocalDateTimeUtils.kt]: We use LocalDateTime in our execution. In addition, I decided to leave it for the future.

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt [hunk 8]
```diff
@@ -98,30 +103,46 @@ class ExecutionController(private val executionService: ExecutionService,
...
-     * @param name
+     * @param projectName
      * @param authentication
      * @param organizationName
+     * @param filters
      * @return list of execution dtos
      */
-    @GetMapping(path = ["/api/$v1/executionDtoList"])
+    @Suppress("PARAMETER_NAME_IN_OUTER_LAMBDA")
+    @PostMapping(path = ["/api/$v1/executionDtoList"])
```
Comment (sanyavertolet) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ExecutionController.kt:113]: Why `POST`? Isn't it more about `GET`?

**Review by sanyavertolet — APPROVED** [review]

**Review by petertrr — APPROVED** [review]
